Merge 0.5

Pbellive opened the pull request:
Hi Lars,
               I've integrated your latest jInv updates and made a couple of changes to make sure it all works on 0.5. All tests are passing. I think each commit represents a discrete piece of work and that they form a logical sequence but I can squash them together if you'd prefer. Thanks for all your input today by the way. It was nice to have that meeting.

Patrick

lruthotto committed:
This is a combination of commits representing changes made to update jInv for compatibility with julia 0.5

Fixed deprecation warnings/errors in Utils and InverseSolve modules. All tests pass for Utils, not sure about InverseSolve.

Finished updating LinearSolvers.jl for 0.5 also fixed method override warnings in InverseSolve. InverseSolve still has other test errors.

Updated getEdgeIntegralOfPolygonalChain to comply with julia 0.5 singleton array dimension handling.

Finished updating for 0.5. Major change was replacing remoteReferences with combination of Future and RemoteChannel. This neccessitated introduction of new helper function initRemoteChannel. Ideally this could be implemented as a macro that works like spawn but returns RemoteChannel instead of Future

Fixed deprecation warning in juliaWrapper.jl

Fixed performance bug in computeMisfit. Local conductivities were being initialized on master process, then copied to workers. They're now initialized on workers

Add docstring for initRemoteChannel

Fixed getData methods in getDataParallel.jl and added tests for them. Deleted old commented out version of nodal diffusion regularization. Fixed parallel loading of modules in setupTests.jl

Brought back doClear to getData methods in getDataParallel.jl and added typing of mesh2mesh input argument. Added doClear field to getData method defined for LSparam forward problem used in tests.

lruthotto commented on src/InverseSolve/computeGradMisfit.jl:
```diff
@@ -80,67 +83,67 @@ function computeGradMisfit(sigma,
 end
 
-# function computeGradMisfit(mc,Dc::Array,pMis::Array{RemoteRef},M2M::Array{RemoteRef},indFors=1:length(PF))
+# function computeGradMisfit(mc,Dc::Array,pMis::Array{Future},M2M::Array{Future},indFors=1:length(PF))
```
this function seems to be commented out. Shall we delete it?

Pbellive replied:
This code seems to date from before the introduction of the misfitParam. I think we can safely delete both this function and lines 50-95 of globalToLocal.jl as they've both become obsolete. Any objections @erantreister?

lruthotto replied:
well not sure this is obsolete. I stumbled over this when running the tests of our EikonalInv package.

Pbellive replied:
Henh. That's interesting, where did it get called from? Not a problem to leave the commented out code there as is. The method seems to be broken as written however. It looks like someone started to update it to use the misfitParam, stopped before they finished and then commented out the code. I suppose the code should either be deleted or fixed and uncommented. Is it meant to be a fault tolerant version of the method just above it?

lruthotto replied:
sorry. At second glance I realize that I probably misplaced my comment here. I think these old methods are not used. Can you see on your machine if EikonalInv tests pass?

Pbellive replied:
All EikonalInv tests pass on my machine after fixing another bug in getDataParallel. Adding getDataParallel tests now.

Pbellive replied:
EikonalInv tests pass on my laptop after fixing up getDataParallel. I didn't delete the commented code but included an additional comment block above the commented out method to explain the reason for keeping it:

```
Method below is old, currently broken and possibly obsolete. However, it seems to include 
functionality for interrupting timed-out workers that is not present anywhere else in jInv. For that 
reason I keep it around but commented out in case anyone wants to resurrect it in the future. 
```

lruthotto commented on src/InverseSolve/regularizers.jl:
```diff
@@ -170,23 +170,23 @@ end
 """
 
-# function wdiffusionRegNodal(m::Vector, mref::Vector, M::AbstractMesh; Iact=1.0, C=[])	
+# function wdiffusionRegNodal(m::Vector, mref::Vector, M::AbstractMesh; Iact=1.0, C=[])
```
shall we delete this function? does not seem to be used.

Pbellive replied:
Seems to have been replaced by a newer method with the same signature. I'll delete it.

lruthotto commented on src/LinearSolvers/blockIterativeWrapper.jl:
```diff
@@ -92,11 +92,11 @@ function solveLinearSystem!(A,B,X,param::BlockIterativeSolver,doTranspose=0)
 			OmInvD = 1./diag(A);
 			Xt      = zeros(n,nrhs)
-			M(R)   = (Xt[:]=0.0; tic(); Xt=ssorPrecTrans!(A,Xt,R,OmInvD); param.timePC+=toq(); return Xt);
+			M = R -> (Xt[:]=0.0; tic(); Xt=ssorPrecTrans!(A,Xt,R,OmInvD); param.timePC+=toq(); return Xt);
```
just curious: Is there a difference between  

```
M(R) = fctn(x1,x2,R) 
```

and

```
M = R-> fctn(x1,x2,R)
```

Pbellive replied:
Yes, in this context there is now an important difference. The problem is that conditionally adding methods to functions doesn't work in 0.5. However, you can get effectively the same behaviour using anonymous functions. See the discussion in the following issue: https://github.com/JuliaLang/julia/issues/15602

lruthotto replied:
Interesting! Do you know if there is a runtime penalty for it

Pbellive replied:
I haven't benchmarked it but there shouldn't be a performance hit. My understanding is that there was a significant performance hit when using anonymous functions prior to 0.5 but now they're pretty much as fast as any other function because the compiler can now infer the result type.

lruthotto commented on src/ForwardShare/getDataParallel.jl:
```diff
@@ -47,21 +47,21 @@ function getData(sigma::Vector,pFor::Array{ForwardProbType},Mesh2Mesh::Array=one
 	workerList = unique(workerList)
 	# prepare remoterefs for sigma
-	sigmaRef = Array(RemoteRef{Channel{Any}},maximum(workers()))
+	sigmaRef = Array(RemoteChannel,maximum(workers()))
```
should this be Futures?

Pbellive replied:
Yes. Thanks for the catch. I'll change that. Shows that getDataParallel.jl isn't being tested. I've put adding those tests on my to do list...along with adding tests to JOcTree.

lruthotto commented on src/ForwardShare/getDataParallel.jl:
```diff
@@ -47,21 +47,21 @@ function getData(sigma::Vector,pFor::Array{ForwardProbType},Mesh2Mesh::Array=one
 	## Compute Dobs
-	Dobs = Array(RemoteRef{Channel{Any}},length(pFor))
-	
+	Dobs = Array(RemoteChannel,length(pFor))
```
should this be Futures?

lruthotto reviewed (commented):
hey patrick. great job, but I ran into some minor questions when trying the code for some of our packages. Can you also double check that the arguments in all  calls to remotecall are  in the new order (function,pid,args...)

Pbellive commented:
I've double checked that all remotecalls use 0.5 syntax (the ones in getDataParallel.jl weren't being captured by the tests), fixed getDataParallel and added tests for it. Should be ready for merging. Do let me know if I've missed anything and we're not ready to merge.

erantreister commented:
Hi,

Patrick - I think that you're right about this code. I don't think it is
relevant anywhere, but let's leave it commented out for now..

I tried to get the EikonalInv tests going privately on my machine and also
had to fix the getDataParallel code. Did you fix something in the
EikonalInv test too? if so - can you push this?

Thanks! Great work!
Eran.

On Fri, Nov 11, 2016 at 7:30 AM, Patrick Belliveau <notifications@github.com

> wrote:
> 
> ## _@Pbellive_ commented on this pull request.
> 
> In src/InverseSolve/computeGradMisfit.jl
> https://github.com/JuliaInv/jInv.jl/pull/23:
> 
> > ```
> > totalTime = toq()
> > ```
> > 
> >   return dF
> >  end
> 
> -# function computeGradMisfit(mc,Dc::Array,pMis::Array{RemoteRef},M2M::Array{RemoteRef},indFors=1:length(PF))
> +# function computeGradMisfit(mc,Dc::Array,pMis::Array{Future},M2M::Array{Future},indFors=1:length(PF))
> 
> EikonalInv tests pass on my laptop after fixing up getDataParallel. I
> didn't delete the commented code but included an additional comment block
> above the commented out method to explain the reason for keeping it:
> 
> Method below is old, currently broken and possibly obsolete. However, it seems to include
> functionality for interrupting timed-out workers that is not present anywhere else in jInv. For that
> reason I keep it around but commented out in case anyone wants to resurrect it in the future.
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> https://github.com/JuliaInv/jInv.jl/pull/23, or mute the thread
> https://github.com/notifications/unsubscribe-auth/AO5HPeEe_19vDVi8htH9uIZF23DYsX8fks5q8_1igaJpZM4KqKNx
> .

erantreister commented:
Hi Patrick & Lars,

I saw that the EikonalInv package was updated but the tests fail. They fail
also on my machine with the updated packages.
The issues that I'm facing are the types in getData - a confusion with the
doClear parameter and M2M parameter.

But first thing:
There used to be a doClear flag in getData which is no longer available in
the methods of getDataParallel.jl.
This is an important flag - determines whether we wish to save sensitivity
information while getting data (factorizations, interpolations etc). Was
this functionality removed or replaced by something else?

Eran.

On Fri, Nov 11, 2016 at 11:28 AM, Eran Treister erantreister@gmail.com
wrote:

> Hi,
> 
> Patrick - I think that you're right about this code. I don't think it is
> relevant anywhere, but let's leave it commented out for now..
> 
> I tried to get the EikonalInv tests going privately on my machine and also
> had to fix the getDataParallel code. Did you fix something in the
> EikonalInv test too? if so - can you push this?
> 
> Thanks! Great work!
> Eran.
> 
> On Fri, Nov 11, 2016 at 7:30 AM, Patrick Belliveau <
> notifications@github.com> wrote:
> 
> > ## _@Pbellive_ commented on this pull request.
> > 
> > In src/InverseSolve/computeGradMisfit.jl
> > https://github.com/JuliaInv/jInv.jl/pull/23:
> > 
> > >    totalTime = toq()
> > >      return dF
> > >  end
> > 
> > -# function computeGradMisfit(mc,Dc::Array,pMis::Array{RemoteRef},M2M::Array{RemoteRef},indFors=1:length(PF))
> > +# function computeGradMisfit(mc,Dc::Array,pMis::Array{Future},M2M::Array{Future},indFors=1:length(PF))
> > 
> > EikonalInv tests pass on my laptop after fixing up getDataParallel. I
> > didn't delete the commented code but included an additional comment block
> > above the commented out method to explain the reason for keeping it:
> > 
> > Method below is old, currently broken and possibly obsolete. However, it seems to include
> > functionality for interrupting timed-out workers that is not present anywhere else in jInv. For that
> > reason I keep it around but commented out in case anyone wants to resurrect it in the future.
> > 
> > —
> > You are receiving this because you were mentioned.
> > Reply to this email directly, view it on GitHub
> > https://github.com/JuliaInv/jInv.jl/pull/23, or mute the thread
> > https://github.com/notifications/unsubscribe-auth/AO5HPeEe_19vDVi8htH9uIZF23DYsX8fks5q8_1igaJpZM4KqKNx
> > .

lruthotto commented:
@Eran: Did you test EikonalInv using Patrick's jInv branch here? For me this works. 
@Patrick: Can you bring the doClear flags back in getDataParallel.jl? It seems to me they went missing during this PR. Also, can you then squash your commits so that we can merge this PR?

erantreister commented:
Hi Lars
Note the name tags: it's @erantreister and @Pbellive. Eran and Patrick are other people...

For me the eikonal test is not working. Maybe it's a windows thing, but it's weird any case. I'm using the current release (see output below - Julia gets confused between the boolean parameter of doClear and the unknown parameter of M2M). Regarding M2M: can we type it? One thing that gets people confusing regarding this package is M2M and globalToLocal. M2M is just the interpolation matrix or a scalar, so I suggest typing it as Union{SparseMatrixCSC,AbstractFloat} - this is also a readability issue. Any other type will not work anyway.

Eran. 

Here is what I get the for eikonal tests:
julia> include("runtests.jl")  
ERROR: LoadError: LoadError: MethodError: jInv.ForwardShare.getData(::Array{Float64,1}, ::EikonalInv.EikonalInvParam, ::Bool) is ambiguous. Candidates:  
  getData(m, pFor::EikonalInv.EikonalInvParam, doClear::Bool) at C:\Users\etrieste\Dropbox\Eran\JuliaCode\Active\EikonalInv.jl\src\getData.jl:6  
  getData(sigma::Array{T<:Any,1}, pFor::jInv.ForwardShare.ForwardProbType, Mesh2Mesh) at C:\Users\etrieste\Dropbox\Eran\JuliaCode\Active\jInv.jl\src\ForwardShare\getDataParallel.jl:21  
 in getData(::Array{Float64,1}, ::EikonalInv.EikonalInvParam) at C:\Users\etrieste\Dropbox\Eran\JuliaCode\Active\EikonalInv.jl\src\getData.jl:6  
 in include_from_node1(::String) at .\loading.jl:488 (repeats 2 times)  
while loading C:\Users\etrieste\Dropbox\Eran\JuliaCode\Active\EikonalInv.jl\test\testGetData2D.jl, in expression starting on line 28  
while loading C:\Users\etrieste\Dropbox\Eran\JuliaCode\Active\EikonalInv.jl\test\runtests.jl, in expression starting on line 1

Pbellive commented:
Hi,
     I've brought back `doClear`, updated the tests of the parallel `getData` methods and added typing of the mesh2mesh input argument. It can be an array of `Future`s, `AbstractFloat`s or sparse matrices. @lruthotto, do you want all this squashed down to 1 commit? All the commits are related to updating for 0.5 but they represent separate chunks of work for the most part. If I squash it down to 1 commit I'll keep the separate commit messages to give a full account of changes.

lruthotto commented:
Thanks for bringing `doClear` back. For me the tests of most packages work now, and I'd like to go ahead and merge this PR. It would be great if you could squash all commits and combine the commit messages so that all changes are reflected. Thanks a lot!

Pbellive committed:
Merged master into 0.5-dev to fix conflicts arising from commit squashing rebase of 0.5 update commits

Pbellive commented:
Alright, I think everything is finally ready to go now. Everything has been squashed into one commit showing the actual work done and a merge commit. The merge commit is necessary because the rebasing (squashing) process scrambled the history on my 0.5-dev branch and master needed to be merged into the development branch one more time for github to recognize that it's up to date.

lruthotto commented:
Thanks, thanks, thank! Great job!
